In [2]:
import pandas as pd
import numpy as np

# Response Variable - Salary

In [3]:
df_s_2021 = pd.read_csv("../dataset/salary/s2021.csv")
df_s_2021.rename(columns={'Unnamed: 1': 'PLAYER', '2021/22': 'SALARY'}, inplace=True)
df_s_2021 = df_s_2021.loc[:,['PLAYER', 'SALARY']]
df_s_2021 = df_s_2021.iloc[0:536,:]
df_s_2021.loc[:,'SEASON'] = 2021
df_s_2021

PLAYER        SALARY  SEASON
0        Stephen Curry  $45,780,966     2021
1         James Harden  $44,310,840     2021
2            John Wall  $44,310,840     2021
3    Russell Westbrook  $44,211,146     2021
4         LeBron James  $41,180,544     2021
..                 ...           ...     ...
531  Admiral Schofield     $300,000     2021
532      Troy Williams     $122,741     2021
533     Mamadi Diakite     $100,000     2021
534  Demetrius Jackson      $92,857     2021
535  Freddie Gillespie      $50,000     2021

[536 rows x 3 columns]

In [4]:
df_s_2020 = pd.read_csv("../dataset/salary/s2020.csv")
df_s_2020.rename(columns={'2020/21(*)': 'SALARY'}, inplace=True)
df_s_2020 = df_s_2020.loc[:,['PLAYER', 'SALARY']]
df_s_2020 = df_s_2020.iloc[0:578,:]
df_s_2020.loc[:,'SEASON'] = 2020
df_s_2020

PLAYER        SALARY  SEASON
0        Stephen Curry  $43,006,362     2020
1    Russell Westbrook  $41,358,814     2020
2           Chris Paul  $41,358,814     2020
3            John Wall  $41,254,920     2020
4         James Harden  $41,254,920     2020
..                 ...           ...     ...
573       Cam Reynolds      $33,299     2020
574       Andrew White      $25,000     2020
575         Tyler Hall      $25,000     2020
576      Elijah Bryant      $24,611     2020
577     Caleb Homesley      $20,000     2020

[578 rows x 3 columns]

In [5]:
df_s_2019 = pd.read_csv("../dataset/salary/s2019.csv")
df_s_2019.rename(columns={'2019/20(*)': 'SALARY'}, inplace=True)
df_s_2019 = df_s_2019.loc[:,['PLAYER', 'SALARY']]
df_s_2019 = df_s_2019.iloc[0:513,:]
df_s_2019.loc[:,'SEASON'] = 2019
df_s_2019

PLAYER        SALARY  SEASON
0        Stephen Curry  $40,231,758     2019
1           Chris Paul  $38,506,482     2019
2    Russell Westbrook  $38,506,482     2019
3            John Wall  $38,199,000     2019
4         James Harden  $38,199,000     2019
..                 ...           ...     ...
508    Jeremiah Martin      $79,568     2019
509     Tremont Waters      $79,568     2019
510         Tacko Fall      $79,568     2019
511      Charlie Brown      $79,568     2019
512       Malik Newman      $65,978     2019

[513 rows x 3 columns]

In [6]:
df_s_2018 = pd.read_csv("../dataset/salary/s2018.csv")
df_s_2018.rename(columns={'2018/19(*)': 'SALARY'}, inplace=True)
df_s_2018 = df_s_2018.loc[:,['PLAYER', 'SALARY']]
df_s_2018 = df_s_2018.iloc[0:576,:]
df_s_2018.loc[:,'SEASON'] = 2018
df_s_2018

PLAYER        SALARY  SEASON
0        Stephen Curry  $37,457,154     2018
1    Russell Westbrook  $35,665,000     2018
2         LeBron James  $35,654,150     2018
3           Chris Paul  $35,654,150     2018
4           Kyle Lowry  $32,700,000     2018
..                 ...           ...     ...
571    Michael Frazier      $47,370     2018
572     Tahjere McCall      $47,370     2018
573     Mitchell Creek      $47,370     2018
574    Isaac Humphries      $47,370     2018
575      Jordan Sibert      $47,370     2018

[576 rows x 3 columns]

In [7]:
df_s_2017 = pd.read_csv("../dataset/salary/s2017.csv")
df_s_2017.rename(columns={'2017/18(*)': 'SALARY'}, inplace=True)
df_s_2017 = df_s_2017.loc[:,['PLAYER', 'SALARY']]
df_s_2017 = df_s_2017.iloc[0:586,:]
df_s_2017.loc[:,'SEASON'] = 2017
df_s_2017

PLAYER        SALARY  SEASON
0          Stephen Curry  $34,682,550     2017
1           LeBron James  $33,285,709     2017
2           Paul Millsap  $30,769,231     2017
3         Gordon Hayward  $29,727,900     2017
4          Blake Griffin  $29,512,900     2017
..                   ...           ...     ...
581         Andre Ingram      $46,079     2017
582  Trey McKinney-Jones      $46,079     2017
583        Aaron Jackson      $46,079     2017
584        Jameel Warney      $46,079     2017
585   Marcus Thornton II      $46,079     2017

[586 rows x 3 columns]

In [8]:
df_s_2016 = pd.read_csv("../dataset/salary/s2016.csv")
df_s_2016.rename(columns={'2016/17(*)': 'SALARY'}, inplace=True)
df_s_2016 = df_s_2016.loc[:,['PLAYER', 'SALARY']]
df_s_2016 = df_s_2016.iloc[0:545,:]
df_s_2016.loc[:,'SEASON'] = 2016
df_s_2016

PLAYER        SALARY  SEASON
0         LeBron James  $30,963,450     2016
1        DeMar DeRozan  $26,540,100     2016
2    Russell Westbrook  $26,540,100     2016
3          Mike Conley  $26,540,100     2016
4         James Harden  $26,540,100     2016
..                 ...           ...     ...
540   Nikola Jovanovic      $30,000     2016
541        Tony Wroten      $25,000     2016
542      Nick Zeisloft      $25,000     2016
543        Jalen Jones      $25,000     2016
544      Dahntay Jones      $24,022     2016

[545 rows x 3 columns]

In [9]:
df_s_2015 = pd.read_csv("../dataset/salary/s2015.csv")
df_s_2015.rename(columns={'2015/16(*)': 'SALARY'}, inplace=True)
df_s_2015 = df_s_2015.loc[:,['PLAYER', 'SALARY']]
df_s_2015 = df_s_2015.iloc[0:500,:]
df_s_2015.loc[:,'SEASON'] = 2015
df_s_2015

PLAYER        SALARY  SEASON
0               Kobe Bryant  $25,000,000     2015
1              LeBron James  $22,970,500     2015
2           Carmelo Anthony  $22,875,000     2015
3             Dwight Howard  $22,359,364     2015
4               Joe Johnson  $22,309,344     2015
..                      ...           ...     ...
495              JJ O'Brien      $30,887     2015
496  Thanasis Antetokounmpo      $30,887     2015
497           Nate Robinson      $26,456     2015
498           Dorell Wright      $17,637     2015
499            John Holland       $9,266     2015

[500 rows x 3 columns]

In [10]:
df_s_2014 = pd.read_csv("../dataset/salary/s2014.csv")
df_s_2014.rename(columns={'2014/15(*)': 'SALARY'}, inplace=True)
df_s_2014 = df_s_2014.loc[:,['PLAYER', 'SALARY']]
df_s_2014 = df_s_2014.iloc[0:513,:]
df_s_2014.loc[:,'SEASON'] = 2014
df_s_2014

PLAYER        SALARY  SEASON
0         Kobe Bryant  $23,500,000     2014
1         Joe Johnson  $23,180,790     2014
2     Carmelo Anthony  $22,458,401     2014
3    Amare Stoudemire  $21,896,658     2014
4       Dwight Howard  $21,436,271     2014
..                ...           ...     ...
508       Sim Bhullar      $35,000     2014
509        David Wear      $29,843     2014
510   Sean Kilpatrick      $29,843     2014
511      Jarell Eddie      $29,843     2014
512        Josh Davis      $20,000     2014

[513 rows x 3 columns]

In [11]:
df_s_2013 = pd.read_csv("../dataset/salary/s2013.csv")
df_s_2013.rename(columns={'2013/14(*)': 'SALARY'}, inplace=True)
df_s_2013 = df_s_2013.loc[:,['PLAYER', 'SALARY']]
df_s_2013 = df_s_2013.iloc[0:492,:]
df_s_2013.loc[:,'SEASON'] = 2013
df_s_2013

PLAYER        SALARY  SEASON
0         Kobe Bryant  $30,453,000     2013
1       Dirk Nowitzki  $22,721,381     2013
2      Gilbert Arenas  $22,346,536     2013
3    Amare Stoudemire  $21,679,893     2013
4     Carmelo Anthony  $21,490,000     2013
..                ...           ...     ...
487    Mustafa Shakur      $46,404     2013
488       DJ Stephens      $28,834     2013
489     Tony Mitchell      $28,834     2013
490     Shane Edwards      $28,834     2013
491       Vander Blue      $28,834     2013

[492 rows x 3 columns]

In [12]:
df_s_2012 = pd.read_csv("../dataset/salary/s2012.csv")
df_s_2012.rename(columns={'2012/13(*)': 'SALARY'}, inplace=True)
df_s_2012 = df_s_2012.loc[:,['PLAYER', 'SALARY']]
df_s_2012 = df_s_2012.iloc[0:494,:]
df_s_2012.loc[:,'SEASON'] = 2012
df_s_2012

PLAYER        SALARY  SEASON
0         Kobe Bryant  $27,849,000     2012
1         Brandon Roy  $21,459,805     2012
2       Dirk Nowitzki  $20,907,128     2012
3      Gilbert Arenas  $20,807,922     2012
4    Amare Stoudemire  $19,948,799     2012
..                ...           ...     ...
489    Justin Dentmon      $44,835     2012
490      Josh Akognon      $41,788     2012
491        Eddy Curry      $31,629     2012
492      Chris Wright      $27,859     2012
493        Henry Sims      $25,073     2012

[494 rows x 3 columns]

In [13]:
df_s_2011 = pd.read_csv("../dataset/salary/s2011.csv")
df_s_2011.rename(columns={'2011/12(*)': 'SALARY'}, inplace=True)
df_s_2011 = df_s_2011.loc[:,['PLAYER', 'SALARY']]
df_s_2011 = df_s_2011.iloc[0:463,:]
df_s_2011.loc[:,'SEASON'] = 2011
df_s_2011

PLAYER        SALARY  SEASON
0         Kobe Bryant  $25,244,493     2011
1        Vince Carter  $21,300,000     2011
2       Kevin Garnett  $21,247,044     2011
3          Tim Duncan  $21,164,619     2011
4       Rashard Lewis  $21,136,631     2011
..                ...           ...     ...
458         Ike Diogu      $64,028     2011
459          Ben Uzoh      $61,433     2011
460      Andre Emmett      $61,433     2011
461      Garret Siler      $51,097     2011
462  Courtney Fortson      $42,009     2011

[463 rows x 3 columns]

In [14]:
frames_s = [df_s_2011, df_s_2012, df_s_2013, df_s_2014, df_s_2015, df_s_2016, df_s_2017, df_s_2018, df_s_2019, df_s_2020]
df_s_2011_2020 = pd.concat(frames_s, axis = 0, join = "inner")
df_s_2011_2020

PLAYER        SALARY  SEASON
0       Kobe Bryant  $25,244,493     2011
1      Vince Carter  $21,300,000     2011
2     Kevin Garnett  $21,247,044     2011
3        Tim Duncan  $21,164,619     2011
4     Rashard Lewis  $21,136,631     2011
..              ...           ...     ...
573    Cam Reynolds      $33,299     2020
574    Andrew White      $25,000     2020
575      Tyler Hall      $25,000     2020
576   Elijah Bryant      $24,611     2020
577  Caleb Homesley      $20,000     2020

[5260 rows x 3 columns]

# Predictors - Traditional + Advanced Data 

In [15]:
df_t_2020 = pd.read_csv("../dataset/player_stat/t2020.csv")
df_t_2020 

Unnamed: 0                 PLAYER TEAM  AGE  GP   W   L   MIN   PTS  \
0             1          Stephen Curry  GSW   33  63  37  26  34.2  32.0   
1             2           Bradley Beal  WAS   28  60  32  28  35.8  31.3   
2             3         Damian Lillard  POR   30  67  39  28  35.8  28.8   
3             4            Joel Embiid  PHI   27  51  39  12  31.1  28.5   
4             5  Giannis Antetokounmpo  MIL   26  61  40  21  33.0  28.1   
..          ...                    ...  ...  ...  ..  ..  ..   ...   ...   
535         536       Anzejs Pasecniks  WAS   25   1   1   0   6.3   0.0   
536         536          Ashton Hagans  MIN   21   2   0   2   2.0   0.0   
537         536       Greg Whittington  DEN   28   4   3   1   3.0   0.0   
538         536            Noah Vonleh  BKN   25   4   3   1   2.7   0.0   
539         536            Will Magnay  NOP   23   1   1   0   3.3   0.0   

      FGM  ...   REB  AST  TOV  STL  BLK   PF    FP  DD2  TD3   +/-  
0    10.4  ...   5.5  5.8  3.4  1.2  0.1  1.9  47.9    8    0   3.9  
1    11.2  ...   4.7  4.4  3.1  1.2  0.4  2.3  45.0    4    0  -0.1  
2     9.0  ...   4.2  7.5  3.0  0.9  0.3  1.5  45.7   16    0   3.0  
3     9.0  ...  10.6  2.8  3.1  1.0  1.4  2.4  49.3   31    0   7.9  
4    10.3  ...  11.0  5.9  3.4  1.2  1.2  2.8  53.9   41    7   6.7  
..    ...  ...   ...  ...  ...  ...  ...  ...   ...  ...  ...   ...  
535   0.0  ...   1.0  1.0  5.0  0.0  0.0  2.0  -2.3    0    0 -13.0  
536   0.0  ...   0.0  0.0  0.5  0.0  0.0  0.0  -0.5    0    0  -1.5  
537   0.0  ...   0.0  0.0  0.0  0.0  0.0  0.0   0.0    0    0   2.3  
538   0.0  ...   0.3  0.3  0.5  0.0  0.0  0.5   0.2    0    0  -4.0  
539   0.0  ...   0.0  0.0  1.0  0.0  0.0  1.0  -1.0    0    0 -10.0  

[540 rows x 30 columns]

In [16]:
df_t_2019 = pd.read_csv("../dataset/player_stat/t2019.csv")
df_t_2018 = pd.read_csv("../dataset/player_stat/t2018.csv")
df_t_2017 = pd.read_csv("../dataset/player_stat/t2017.csv")
df_t_2016 = pd.read_csv("../dataset/player_stat/t2016.csv")
df_t_2015 = pd.read_csv("../dataset/player_stat/t2015.csv")
df_t_2014 = pd.read_csv("../dataset/player_stat/t2014.csv")
df_t_2013 = pd.read_csv("../dataset/player_stat/t2013.csv")
df_t_2012 = pd.read_csv("../dataset/player_stat/t2012.csv")
df_t_2011 = pd.read_csv("../dataset/player_stat/t2011.csv")
df_t_2010 = pd.read_csv("../dataset/player_stat/t2010.csv")

In [17]:
df_t_2010
# df_t_2020.columns

Unnamed: 0            PLAYER TEAM  AGE  GP   W   L   MIN   PTS  FGM  ...  \
0             1      Kevin Durant  OKC   22  78  52  26  38.9  27.7  9.1  ...   
1             2      LeBron James  MIA   26  79  57  22  38.8  26.7  9.6  ...   
2             3   Carmelo Anthony  NYK   27  77  43  34  35.7  25.6  8.9  ...   
3             4       Dwyane Wade  MIA   29  76  54  22  37.1  25.5  9.1  ...   
4             5       Kobe Bryant  LAL   32  82  57  25  33.9  25.3  9.0  ...   
..          ...               ...  ...  ...  ..  ..  ..   ...   ...  ...  ...   
447         448      Theo Ratliff  LAL   38  10  10   0   7.1   0.2  0.1  ...   
448         449     Bobby Simmons  SAS   31   2   2   0   7.8   0.0  0.0  ...   
449         449     Brian Skinner  MIL   35   2   2   0   2.9   0.0  0.0  ...   
450         449        Gani Lawal  PHX   22   1   1   0   1.9   0.0  0.0  ...   
451         449  Hassan Whiteside  SAC   22   1   1   0   1.7   0.0  0.0  ...   

     REB  AST  TOV  STL  BLK   PF    FP  DD2  TD3  +/-  
0    6.8  2.7  2.8  1.1  1.0  2.0  43.5   14    0  3.4  
1    7.5  7.0  3.6  1.6  0.6  2.1  49.2   31    4  7.8  
2    7.3  2.9  2.7  0.9  0.6  2.9  40.4   18    0  2.4  
3    6.4  4.6  3.1  1.5  1.1  2.6  44.7   10    1  7.6  
4    5.1  4.7  3.0  1.2  0.1  2.1  39.7    9    1  5.2  
..   ...  ...  ...  ...  ...  ...   ...  ...  ...  ...  
447  1.3  0.3  0.3  0.2  0.5  0.7   4.0    0    0 -2.7  
448  0.0  1.0  0.0  0.0  0.0  1.0   1.5    0    0  3.0  
449  0.0  0.0  0.0  0.0  0.0  0.0   0.0    0    0 -5.5  
450  0.0  0.0  0.0  0.0  0.0  1.0   0.0    0    0 -2.0  
451  0.0  0.0  0.0  0.0  0.0  2.0   0.0    0    0 -4.0  

[452 rows x 30 columns]

In [18]:
df_a_2020 = pd.read_csv("../dataset/player_stat/a2020.csv")
df_a_2019 = pd.read_csv("../dataset/player_stat/a2019.csv")
df_a_2018 = pd.read_csv("../dataset/player_stat/a2018.csv")
df_a_2017 = pd.read_csv("../dataset/player_stat/a2017.csv")
df_a_2016 = pd.read_csv("../dataset/player_stat/a2016.csv")
df_a_2015 = pd.read_csv("../dataset/player_stat/a2015.csv")
df_a_2014 = pd.read_csv("../dataset/player_stat/a2014.csv")
df_a_2013 = pd.read_csv("../dataset/player_stat/a2013.csv")
df_a_2012 = pd.read_csv("../dataset/player_stat/a2012.csv")
df_a_2011 = pd.read_csv("../dataset/player_stat/a2011.csv")
df_a_2010 = pd.read_csv("../dataset/player_stat/a2010.csv")

In [19]:
df_a_2010
# df_a_2020.columns

Unnamed: 0            PLAYER TEAM  AGE  GP   W   L   MIN  OFFRTG  DEFRTG  \
0             1      Jarrett Jack  NOH   27  83  39  44  20.8   101.6   106.3   
1             2      Al Jefferson  UTA   26  82  39  43  35.9   107.7   110.8   
2             2        Ben Gordon  DET   28  82  30  52  26.0   107.2   110.0   
3             2     Blake Griffin  LAC   22  82  32  50  37.9   105.9   108.0   
4             2        Boris Diaw  CHA   29  82  34  48  33.9   102.8   107.3   
..          ...               ...  ...  ...  ..  ..  ..   ...     ...     ...   
447         447     Brian Skinner  MIL   35   2   2   0   2.9    57.1   146.2   
448         447    Dexter Pittman  MIA   23   2   2   0   5.6   136.4    95.7   
449         447    John Lucas III  CHI   28   2   1   1   5.1    89.5   105.3   
450         451        Gani Lawal  PHX   22   1   1   0   1.9    50.0   100.0   
451         451  Hassan Whiteside  SAC   22   1   1   0   1.7     0.0   133.3   

     ...  OREB%  DREB%  REB%  TO RATIO  EFG%   TS%  USG%    PACE    PIE   POSS  
0    ...    1.2    9.6   5.3      10.2  43.4  50.3  22.4   90.61   10.9  3,239  
1    ...    8.5   20.2  14.3       6.2  49.6  52.8  23.9   92.19   13.4  5,651  
2    ...    2.0    8.2   4.9      12.4  51.0  54.8  20.6   91.49    7.5  4,061  
3    ...    9.0   23.8  16.4      10.1  50.8  54.9  27.0   93.88   16.0  6,078  
4    ...    4.0   11.8   7.9      12.5  54.3  55.8  16.6   90.97   10.0  5,265  
..   ...    ...    ...   ...       ...   ...   ...   ...     ...    ...    ...  
447  ...    0.0    0.0   0.0       0.0   0.0   0.0   0.0  110.14    0.0     14  
448  ...   12.5   16.7  15.0      40.0  33.3  33.3  20.0   96.14   -2.6     22  
449  ...    0.0    0.0   0.0       0.0  33.3  25.8  19.0   90.24   -5.3     19  
450  ...    0.0    0.0   0.0       0.0   0.0   0.0   0.0   99.31  -40.0      4  
451  ...    0.0    0.0   0.0       0.0   0.0   0.0   0.0   82.60 -200.0      3  

[452 rows x 24 columns]

In [20]:
df_ta_2020 = pd.merge(df_t_2020, df_a_2020, on = ['PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN'], how = 'inner')
df_ta_2020

Unnamed: 0_x                 PLAYER TEAM  AGE  GP   W   L   MIN   PTS  \
0               1          Stephen Curry  GSW   33  63  37  26  34.2  32.0   
1               2           Bradley Beal  WAS   28  60  32  28  35.8  31.3   
2               3         Damian Lillard  POR   30  67  39  28  35.8  28.8   
3               4            Joel Embiid  PHI   27  51  39  12  31.1  28.5   
4               5  Giannis Antetokounmpo  MIL   26  61  40  21  33.0  28.1   
..            ...                    ...  ...  ...  ..  ..  ..   ...   ...   
535           536       Anzejs Pasecniks  WAS   25   1   1   0   6.3   0.0   
536           536          Ashton Hagans  MIN   21   2   0   2   2.0   0.0   
537           536       Greg Whittington  DEN   28   4   3   1   3.0   0.0   
538           536            Noah Vonleh  BKN   25   4   3   1   2.7   0.0   
539           536            Will Magnay  NOP   23   1   1   0   3.3   0.0   

      FGM  ...  OREB%  DREB%  REB%  TO RATIO  EFG%   TS%  USG%    PACE   PIE  \
0    10.4  ...    1.3   13.5   7.7      10.2  60.5  65.5  33.1  105.16  18.1   
1    11.2  ...    3.3    9.4   6.4       9.3  53.2  59.3  33.1  105.21  14.5   
2     9.0  ...    1.2   10.4   5.7       9.0  55.4  62.3  30.6  100.40  15.8   
3     9.0  ...    7.8   25.4  17.2      11.0  54.5  63.6  35.2  100.79  20.3   
4    10.3  ...    4.8   26.4  16.0      10.8  60.0  63.3  32.0  104.17  19.6   
..    ...  ...    ...    ...   ...       ...   ...   ...   ...     ...   ...   
535   0.0  ...   14.3    0.0   8.3      71.4   0.0   0.0  37.5  102.59 -26.0   
536   0.0  ...    0.0    0.0   0.0     100.0   0.0   0.0   8.3  123.93 -25.0   
537   0.0  ...    0.0    0.0   0.0       0.0   0.0   0.0  10.3  107.85  -7.1   
538   0.0  ...    0.0   14.3   6.7      33.3   0.0   0.0  21.7  104.77  -9.4   
539   0.0  ...    0.0    0.0   0.0      50.0   0.0   0.0  20.0  110.20 -17.1   

      POSS  
0    4,725  
1    4,713  
2    5,014  
3    3,331  
4    4,359  
..     ...  
535     14  
536     10  
537     27  
538     23  
539      8  

[540 rows x 47 columns]

In [21]:
df_ta_2019 = pd.merge(df_t_2019, df_a_2019, on = ['PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN'], how = 'inner')
df_ta_2018 = pd.merge(df_t_2018, df_a_2018, on = ['PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN'], how = 'inner')
df_ta_2017 = pd.merge(df_t_2017, df_a_2017, on = ['PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN'], how = 'inner')
df_ta_2016 = pd.merge(df_t_2016, df_a_2016, on = ['PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN'], how = 'inner')
df_ta_2015 = pd.merge(df_t_2015, df_a_2015, on = ['PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN'], how = 'inner')
df_ta_2014 = pd.merge(df_t_2014, df_a_2014, on = ['PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN'], how = 'inner')
df_ta_2013 = pd.merge(df_t_2013, df_a_2013, on = ['PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN'], how = 'inner')
df_ta_2012 = pd.merge(df_t_2012, df_a_2012, on = ['PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN'], how = 'inner')
df_ta_2011 = pd.merge(df_t_2011, df_a_2011, on = ['PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN'], how = 'inner')
df_ta_2010 = pd.merge(df_t_2010, df_a_2010, on = ['PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN'], how = 'inner')

In [22]:
print(len(df_t_2010),len(df_a_2010))
df_ta_2010.columns

452 452


Index(['Unnamed: 0_x', 'PLAYER', 'TEAM', 'AGE', 'GP', 'W', 'L', 'MIN', 'PTS',
       'FGM', 'FGA', 'FG%', '3PM', '3PA', '3P%', 'FTM', 'FTA', 'FT%', 'OREB',
       'DREB', 'REB', 'AST', 'TOV', 'STL', 'BLK', 'PF', 'FP', 'DD2', 'TD3',
       '+/-', 'Unnamed: 0_y', 'OFFRTG', 'DEFRTG', 'NETRTG', 'AST%', 'AST/TO',
       'AST RATIO', 'OREB%', 'DREB%', 'REB%', 'TO RATIO', 'EFG%', 'TS%',
       'USG%', 'PACE', 'PIE', 'POSS'],
      dtype='object')

In [23]:
df_ta_2010.loc[:, 'SEASON'] = 2010
df_ta_2011.loc[:, 'SEASON'] = 2011
df_ta_2012.loc[:, 'SEASON'] = 2012
df_ta_2013.loc[:, 'SEASON'] = 2013
df_ta_2014.loc[:, 'SEASON'] = 2014
df_ta_2015.loc[:, 'SEASON'] = 2015
df_ta_2016.loc[:, 'SEASON'] = 2016
df_ta_2017.loc[:, 'SEASON'] = 2017
df_ta_2018.loc[:, 'SEASON'] = 2018
df_ta_2019.loc[:, 'SEASON'] = 2019
df_ta_2020.loc[:, 'SEASON'] = 2020
df_ta_2010

Unnamed: 0_x            PLAYER TEAM  AGE  GP   W   L   MIN   PTS  FGM  \
0               1      Kevin Durant  OKC   22  78  52  26  38.9  27.7  9.1   
1               2      LeBron James  MIA   26  79  57  22  38.8  26.7  9.6   
2               3   Carmelo Anthony  NYK   27  77  43  34  35.7  25.6  8.9   
3               4       Dwyane Wade  MIA   29  76  54  22  37.1  25.5  9.1   
4               5       Kobe Bryant  LAL   32  82  57  25  33.9  25.3  9.0   
..            ...               ...  ...  ...  ..  ..  ..   ...   ...  ...   
447           448      Theo Ratliff  LAL   38  10  10   0   7.1   0.2  0.1   
448           449     Bobby Simmons  SAS   31   2   2   0   7.8   0.0  0.0   
449           449     Brian Skinner  MIL   35   2   2   0   2.9   0.0  0.0   
450           449        Gani Lawal  PHX   22   1   1   0   1.9   0.0  0.0   
451           449  Hassan Whiteside  SAC   22   1   1   0   1.7   0.0  0.0   

     ...  DREB%  REB%  TO RATIO  EFG%   TS%  USG%    PACE    PIE   POSS  \
0    ...   15.4   8.9       9.7  50.9  58.9  29.6   95.02   15.9  6,025   
1    ...   16.3  10.1      10.9  54.1  59.4  30.7   92.52   20.1  5,916   
2    ...   16.0  10.4       9.4  48.7  55.7  31.8   96.24   14.3  5,506   
3    ...   12.7   9.1      10.6  52.3  58.1  30.8   92.97   16.4  5,471   
4    ...   12.1   7.7       9.7  48.7  54.8  34.3   92.36   15.6  5,361   
..   ...    ...   ...       ...   ...   ...   ...     ...    ...    ...   
447  ...    9.5   8.3      25.0  16.7  14.5   5.1  101.10    1.0    146   
448  ...    0.0   0.0       0.0   0.0   0.0   8.6   92.79   -5.8     30   
449  ...    0.0   0.0       0.0   0.0   0.0   0.0  110.14    0.0     14   
450  ...    0.0   0.0       0.0   0.0   0.0   0.0   99.31  -40.0      4   
451  ...    0.0   0.0       0.0   0.0   0.0   0.0   82.60 -200.0      3   

     SEASON  
0      2010  
1      2010  
2      2010  
3      2010  
4      2010  
..      ...  
447    2010  
448    2010  
449    2010  
450    2010  
451    2010  

[452 rows x 48 columns]

In [24]:
frames = [df_ta_2010, df_ta_2011, df_ta_2012, df_ta_2013, df_ta_2014, df_ta_2015, df_ta_2016, df_ta_2017, df_ta_2018, df_ta_2019]
df_ta_2010_2019 = pd.concat(frames, axis = 0, join = "inner")
df_ta_2010_2019


Unnamed: 0_x           PLAYER TEAM  AGE  GP   W   L   MIN   PTS  FGM  \
0               1     Kevin Durant  OKC   22  78  52  26  38.9  27.7  9.1   
1               2     LeBron James  MIA   26  79  57  22  38.8  26.7  9.6   
2               3  Carmelo Anthony  NYK   27  77  43  34  35.7  25.6  8.9   
3               4      Dwyane Wade  MIA   29  76  54  22  37.1  25.5  9.1   
4               5      Kobe Bryant  LAL   32  82  57  25  33.9  25.3  9.0   
..            ...              ...  ...  ...  ..  ..  ..   ...   ...  ...   
524           525      Isaiah Roby  OKC   22   3   2   1   3.8   0.0  0.0   
525           525      J.P. Macura  CLE   25   1   1   0   0.6   0.0  0.0   
526           525   Marques Bolden  CLE   22   1   0   1   2.8   0.0  0.0   
527           525     Stanton Kidd  UTA   28   4   3   1   3.8   0.0  0.0   
528           525   William Howard  HOU   26   2   1   1   6.6   0.0  0.0   

     ...  DREB%  REB%  TO RATIO  EFG%   TS%  USG%    PACE   PIE   POSS  SEASON  
0    ...   15.4   8.9       9.7  50.9  58.9  29.6   95.02  15.9  6,025    2010  
1    ...   16.3  10.1      10.9  54.1  59.4  30.7   92.52  20.1  5,916    2010  
2    ...   16.0  10.4       9.4  48.7  55.7  31.8   96.24  14.3  5,506    2010  
3    ...   12.7   9.1      10.6  52.3  58.1  30.8   92.97  16.4  5,471    2010  
4    ...   12.1   7.7       9.7  48.7  54.8  34.3   92.36  15.6  5,361    2010  
..   ...    ...   ...       ...   ...   ...   ...     ...   ...    ...     ...  
524  ...   16.7   8.0      50.0   0.0   0.0   7.1  111.58  -2.0     26    2019  
525  ...    0.0   0.0       0.0   0.0   0.0   0.0  115.51   0.0      2    2019  
526  ...   50.0  33.3       0.0   0.0   0.0   0.0  119.29  30.8      7    2019  
527  ...   15.8   9.1      33.3   0.0   0.0  13.9  109.92  -7.7     34    2019  
528  ...   25.0   7.7      16.7   0.0   0.0  15.2  106.73  -8.6     29    2019  

[4934 rows x 48 columns]

In [25]:
df_s_2011_2020.loc[:,'SEASON'] = df_s_2011_2020.loc[:,'SEASON'] - 1
df_s_2011_2020

PLAYER        SALARY  SEASON
0       Kobe Bryant  $25,244,493     2010
1      Vince Carter  $21,300,000     2010
2     Kevin Garnett  $21,247,044     2010
3        Tim Duncan  $21,164,619     2010
4     Rashard Lewis  $21,136,631     2010
..              ...           ...     ...
573    Cam Reynolds      $33,299     2019
574    Andrew White      $25,000     2019
575      Tyler Hall      $25,000     2019
576   Elijah Bryant      $24,611     2019
577  Caleb Homesley      $20,000     2019

[5260 rows x 3 columns]

In [26]:
df_ta_2010_2019

Unnamed: 0_x           PLAYER TEAM  AGE  GP   W   L   MIN   PTS  FGM  \
0               1     Kevin Durant  OKC   22  78  52  26  38.9  27.7  9.1   
1               2     LeBron James  MIA   26  79  57  22  38.8  26.7  9.6   
2               3  Carmelo Anthony  NYK   27  77  43  34  35.7  25.6  8.9   
3               4      Dwyane Wade  MIA   29  76  54  22  37.1  25.5  9.1   
4               5      Kobe Bryant  LAL   32  82  57  25  33.9  25.3  9.0   
..            ...              ...  ...  ...  ..  ..  ..   ...   ...  ...   
524           525      Isaiah Roby  OKC   22   3   2   1   3.8   0.0  0.0   
525           525      J.P. Macura  CLE   25   1   1   0   0.6   0.0  0.0   
526           525   Marques Bolden  CLE   22   1   0   1   2.8   0.0  0.0   
527           525     Stanton Kidd  UTA   28   4   3   1   3.8   0.0  0.0   
528           525   William Howard  HOU   26   2   1   1   6.6   0.0  0.0   

     ...  DREB%  REB%  TO RATIO  EFG%   TS%  USG%    PACE   PIE   POSS  SEASON  
0    ...   15.4   8.9       9.7  50.9  58.9  29.6   95.02  15.9  6,025    2010  
1    ...   16.3  10.1      10.9  54.1  59.4  30.7   92.52  20.1  5,916    2010  
2    ...   16.0  10.4       9.4  48.7  55.7  31.8   96.24  14.3  5,506    2010  
3    ...   12.7   9.1      10.6  52.3  58.1  30.8   92.97  16.4  5,471    2010  
4    ...   12.1   7.7       9.7  48.7  54.8  34.3   92.36  15.6  5,361    2010  
..   ...    ...   ...       ...   ...   ...   ...     ...   ...    ...     ...  
524  ...   16.7   8.0      50.0   0.0   0.0   7.1  111.58  -2.0     26    2019  
525  ...    0.0   0.0       0.0   0.0   0.0   0.0  115.51   0.0      2    2019  
526  ...   50.0  33.3       0.0   0.0   0.0   0.0  119.29  30.8      7    2019  
527  ...   15.8   9.1      33.3   0.0   0.0  13.9  109.92  -7.7     34    2019  
528  ...   25.0   7.7      16.7   0.0   0.0  15.2  106.73  -8.6     29    2019  

[4934 rows x 48 columns]

In [27]:
df_sta_2011_2020 = pd.merge(df_s_2011_2020, df_ta_2010_2019, on=['PLAYER','SEASON'], how = 'inner')
# df_sta_2011_2020.isnull().sum()
df_sta_2011_2020

PLAYER        SALARY  SEASON  Unnamed: 0_x TEAM  AGE  GP   W  \
0         Kobe Bryant  $25,244,493     2010             5  LAL   32  82  57   
1        Vince Carter  $21,300,000     2010            74  PHX   34  73  38   
2       Kevin Garnett  $21,247,044     2010            63  BOS   35  71  49   
3          Tim Duncan  $21,164,619     2010            82  SAS   35  76  60   
4       Rashard Lewis  $21,136,631     2010           110  WAS   31  57  24   
...               ...           ...     ...           ...  ...  ...  ..  ..   
3830     Jaylen Hoard     $129,209     2019           441  POR   21  13   7   
3831  Justin Anderson     $125,000     2019           447  BKN   26  10   6   
3832  Jeremiah Martin      $58,452     2019           263  BKN   24   9   6   
3833     Jerian Grant      $50,000     2019           379  WAS   27   6   1   
3834       Louis King      $49,222     2019           475  DET   21  10   3   

       L   MIN  ...  OREB%  DREB%  REB%  TO RATIO  EFG%   TS%  USG%    PACE  \
0     25  33.9  ...    3.1   12.1   7.7       9.7  48.7  54.8  34.3   92.36   
1     35  28.1  ...    2.7   10.8   6.8       7.5  50.2  52.8  23.0   94.68   
2     22  31.3  ...    4.4   24.9  15.5       9.5  52.9  57.5  21.9   92.45   
3     16  28.4  ...    8.4   23.3  16.2       9.5  50.0  53.7  22.5   94.52   
4     33  32.0  ...    3.9   12.2   8.0      10.8  50.8  53.5  17.6   94.05   
...   ..   ...  ...    ...    ...   ...       ...   ...   ...   ...     ...   
3830   6   7.9  ...   11.1   15.8  13.3       8.9  46.9  50.4  15.3  103.22   
3831   4  10.7  ...    0.8   16.7   8.8       7.8  34.2  35.2  16.2  105.32   
3832   3  11.0  ...    1.0    8.9   4.9       4.9  50.0  54.1  26.8  104.38   
3833   5  13.4  ...    2.2    5.6   3.6       7.1  40.7  44.9  16.6  103.05   
3834   7   6.2  ...    1.5   14.1   7.6       6.9  47.6  44.8  16.3   96.85   

       PIE   POSS  
0     15.6  5,361  
1      9.2  4,042  
2     16.8  4,276  
3     15.4  4,261  
4      8.1  3,571  
...    ...    ...  
3830   8.8    223  
3831   2.8    237  
3832  13.3    216  
3833   3.2    174  
3834   4.3    128  

[3835 rows x 49 columns]

In [28]:
df_sta_2011_2020.SALARY = [x[1:] for x in df_sta_2011_2020.SALARY]
df_sta_2011_2020.SALARY = df_sta_2011_2020.SALARY.str.replace(",", "")
df_sta_2011_2020['SALARY'] = pd.to_numeric(df_sta_2011_2020['SALARY'])
df_sta_2011_2020


PLAYER    SALARY  SEASON  Unnamed: 0_x TEAM  AGE  GP   W   L  \
0         Kobe Bryant  25244493    2010             5  LAL   32  82  57  25   
1        Vince Carter  21300000    2010            74  PHX   34  73  38  35   
2       Kevin Garnett  21247044    2010            63  BOS   35  71  49  22   
3          Tim Duncan  21164619    2010            82  SAS   35  76  60  16   
4       Rashard Lewis  21136631    2010           110  WAS   31  57  24  33   
...               ...       ...     ...           ...  ...  ...  ..  ..  ..   
3830     Jaylen Hoard    129209    2019           441  POR   21  13   7   6   
3831  Justin Anderson    125000    2019           447  BKN   26  10   6   4   
3832  Jeremiah Martin     58452    2019           263  BKN   24   9   6   3   
3833     Jerian Grant     50000    2019           379  WAS   27   6   1   5   
3834       Louis King     49222    2019           475  DET   21  10   3   7   

       MIN  ...  OREB%  DREB%  REB%  TO RATIO  EFG%   TS%  USG%    PACE   PIE  \
0     33.9  ...    3.1   12.1   7.7       9.7  48.7  54.8  34.3   92.36  15.6   
1     28.1  ...    2.7   10.8   6.8       7.5  50.2  52.8  23.0   94.68   9.2   
2     31.3  ...    4.4   24.9  15.5       9.5  52.9  57.5  21.9   92.45  16.8   
3     28.4  ...    8.4   23.3  16.2       9.5  50.0  53.7  22.5   94.52  15.4   
4     32.0  ...    3.9   12.2   8.0      10.8  50.8  53.5  17.6   94.05   8.1   
...    ...  ...    ...    ...   ...       ...   ...   ...   ...     ...   ...   
3830   7.9  ...   11.1   15.8  13.3       8.9  46.9  50.4  15.3  103.22   8.8   
3831  10.7  ...    0.8   16.7   8.8       7.8  34.2  35.2  16.2  105.32   2.8   
3832  11.0  ...    1.0    8.9   4.9       4.9  50.0  54.1  26.8  104.38  13.3   
3833  13.4  ...    2.2    5.6   3.6       7.1  40.7  44.9  16.6  103.05   3.2   
3834   6.2  ...    1.5   14.1   7.6       6.9  47.6  44.8  16.3   96.85   4.3   

       POSS  
0     5,361  
1     4,042  
2     4,276  
3     4,261  
4     3,571  
...     ...  
3830    223  
3831    237  
3832    216  
3833    174  
3834    128  

[3835 rows x 49 columns]

In [29]:
df_sta_2011_2020.columns
# df_sta_2011_2020.isnull().sum()
predictors = ['PLAYER', 'SALARY', 'SEASON', 'TEAM', 'AGE', 'GP', 'W', 'L', \
              'PTS', '3PM', 'REB', 'AST', 'TOV', 'STL', 'BLK', '+/-', \
              'OFFRTG', 'DEFRTG', 'NETRTG']
dfPlayerValue = df_sta_2011_2020.loc[:, predictors]
dfPlayerValue

PLAYER    SALARY  SEASON TEAM  AGE  GP   W   L   PTS  3PM  REB  \
0         Kobe Bryant  25244493    2010  LAL   32  82  57  25  25.3  1.4  5.1   
1        Vince Carter  21300000    2010  PHX   34  73  38  35  14.0  1.6  3.8   
2       Kevin Garnett  21247044    2010  BOS   35  71  49  22  14.9  0.0  8.9   
3          Tim Duncan  21164619    2010  SAS   35  76  60  16  13.4  0.0  8.9   
4       Rashard Lewis  21136631    2010  WAS   31  57  24  33  11.7  1.5  5.1   
...               ...       ...     ...  ...  ...  ..  ..  ..   ...  ...  ...   
3830     Jaylen Hoard    129209    2019  POR   21  13   7   6   2.9  0.0  2.5   
3831  Justin Anderson    125000    2019  BKN   26  10   6   4   2.8  0.6  2.1   
3832  Jeremiah Martin     58452    2019  BKN   24   9   6   3   7.1  0.6  1.1   
3833     Jerian Grant     50000    2019  WAS   27   6   1   5   4.5  0.3  1.0   
3834       Louis King     49222    2019  DET   21  10   3   7   2.0  0.4  1.0   

      AST  TOV  STL  BLK  +/-  OFFRTG  DEFRTG  NETRTG  
0     4.7  3.0  1.2  0.1  5.2   111.5   104.2     7.3  
1     2.0  1.2  0.9  0.3  1.7   108.0   104.6     3.4  
2     2.4  1.6  1.3  0.8  7.8   109.2    96.4    12.9  
3     2.7  1.6  0.7  1.9  6.0   110.3   100.3    10.0  
4     1.6  1.5  0.9  0.5 -1.2   102.6   104.3    -1.6  
...   ...  ...  ...  ...  ...     ...     ...     ...  
3830  0.3  0.3  0.4  0.0  0.8   112.1   109.1     3.0  
3831  0.8  0.4  0.0  0.6  0.3   106.8   106.8    -0.1  
3832  2.0  0.4  0.8  0.2 -1.9   107.4   115.8    -8.4  
3833  1.5  0.5  0.2  0.2 -0.5   103.4   107.0    -3.6  
3834  0.5  0.2  0.2  0.0  1.9   109.4    98.4    11.0  

[3835 rows x 19 columns]

In [30]:
df_s_cap = pd.read_csv("../dataset/salary/s_cap.csv")
df_s_cap.SALARY_CAP = [x[1:] for x in df_s_cap.SALARY_CAP]
df_s_cap.SALARY_CAP = df_s_cap.SALARY_CAP.str.replace(",", "")
df_s_cap['SALARY_CAP'] = pd.to_numeric(df_s_cap['SALARY_CAP'])
df_s_cap

SEASON  SALARY_CAP
0     2011    58044000
1     2012    58044000
2     2013    58679000
3     2014    63065000
4     2015    70000000
5     2016    94143000
6     2017    99093000
7     2018   101869000
8     2019   109140000
9     2020   109140000
10    2021   112414000

In [31]:
df_s_cap.loc[:,'SEASON'] = df_s_cap.loc[:,'SEASON'] - 1
df_s_cap

SEASON  SALARY_CAP
0     2010    58044000
1     2011    58044000
2     2012    58679000
3     2013    63065000
4     2014    70000000
5     2015    94143000
6     2016    99093000
7     2017   101869000
8     2018   109140000
9     2019   109140000
10    2020   112414000

In [32]:
dfPlayerValue = pd.merge(dfPlayerValue, df_s_cap, on="SEASON", how = 'outer')
dfPlayerValue

PLAYER      SALARY  SEASON TEAM   AGE    GP     W     L   PTS  \
0         Kobe Bryant  25244493.0    2010  LAL  32.0  82.0  57.0  25.0  25.3   
1        Vince Carter  21300000.0    2010  PHX  34.0  73.0  38.0  35.0  14.0   
2       Kevin Garnett  21247044.0    2010  BOS  35.0  71.0  49.0  22.0  14.9   
3          Tim Duncan  21164619.0    2010  SAS  35.0  76.0  60.0  16.0  13.4   
4       Rashard Lewis  21136631.0    2010  WAS  31.0  57.0  24.0  33.0  11.7   
...               ...         ...     ...  ...   ...   ...   ...   ...   ...   
3831  Justin Anderson    125000.0    2019  BKN  26.0  10.0   6.0   4.0   2.8   
3832  Jeremiah Martin     58452.0    2019  BKN  24.0   9.0   6.0   3.0   7.1   
3833     Jerian Grant     50000.0    2019  WAS  27.0   6.0   1.0   5.0   4.5   
3834       Louis King     49222.0    2019  DET  21.0  10.0   3.0   7.0   2.0   
3835              NaN         NaN    2020  NaN   NaN   NaN   NaN   NaN   NaN   

      3PM  REB  AST  TOV  STL  BLK  +/-  OFFRTG  DEFRTG  NETRTG  SALARY_CAP  
0     1.4  5.1  4.7  3.0  1.2  0.1  5.2   111.5   104.2     7.3    58044000  
1     1.6  3.8  2.0  1.2  0.9  0.3  1.7   108.0   104.6     3.4    58044000  
2     0.0  8.9  2.4  1.6  1.3  0.8  7.8   109.2    96.4    12.9    58044000  
3     0.0  8.9  2.7  1.6  0.7  1.9  6.0   110.3   100.3    10.0    58044000  
4     1.5  5.1  1.6  1.5  0.9  0.5 -1.2   102.6   104.3    -1.6    58044000  
...   ...  ...  ...  ...  ...  ...  ...     ...     ...     ...         ...  
3831  0.6  2.1  0.8  0.4  0.0  0.6  0.3   106.8   106.8    -0.1   109140000  
3832  0.6  1.1  2.0  0.4  0.8  0.2 -1.9   107.4   115.8    -8.4   109140000  
3833  0.3  1.0  1.5  0.5  0.2  0.2 -0.5   103.4   107.0    -3.6   109140000  
3834  0.4  1.0  0.5  0.2  0.2  0.0  1.9   109.4    98.4    11.0   109140000  
3835  NaN  NaN  NaN  NaN  NaN  NaN  NaN     NaN     NaN     NaN   112414000  

[3836 rows x 20 columns]

In [33]:
dfPlayerValue = dfPlayerValue.iloc[0:3835, :]
dfPlayerValue.SALARY_CAP.value_counts()
dfPlayerValue

PLAYER      SALARY  SEASON TEAM   AGE    GP     W     L   PTS  \
0         Kobe Bryant  25244493.0    2010  LAL  32.0  82.0  57.0  25.0  25.3   
1        Vince Carter  21300000.0    2010  PHX  34.0  73.0  38.0  35.0  14.0   
2       Kevin Garnett  21247044.0    2010  BOS  35.0  71.0  49.0  22.0  14.9   
3          Tim Duncan  21164619.0    2010  SAS  35.0  76.0  60.0  16.0  13.4   
4       Rashard Lewis  21136631.0    2010  WAS  31.0  57.0  24.0  33.0  11.7   
...               ...         ...     ...  ...   ...   ...   ...   ...   ...   
3830     Jaylen Hoard    129209.0    2019  POR  21.0  13.0   7.0   6.0   2.9   
3831  Justin Anderson    125000.0    2019  BKN  26.0  10.0   6.0   4.0   2.8   
3832  Jeremiah Martin     58452.0    2019  BKN  24.0   9.0   6.0   3.0   7.1   
3833     Jerian Grant     50000.0    2019  WAS  27.0   6.0   1.0   5.0   4.5   
3834       Louis King     49222.0    2019  DET  21.0  10.0   3.0   7.0   2.0   

      3PM  REB  AST  TOV  STL  BLK  +/-  OFFRTG  DEFRTG  NETRTG  SALARY_CAP  
0     1.4  5.1  4.7  3.0  1.2  0.1  5.2   111.5   104.2     7.3    58044000  
1     1.6  3.8  2.0  1.2  0.9  0.3  1.7   108.0   104.6     3.4    58044000  
2     0.0  8.9  2.4  1.6  1.3  0.8  7.8   109.2    96.4    12.9    58044000  
3     0.0  8.9  2.7  1.6  0.7  1.9  6.0   110.3   100.3    10.0    58044000  
4     1.5  5.1  1.6  1.5  0.9  0.5 -1.2   102.6   104.3    -1.6    58044000  
...   ...  ...  ...  ...  ...  ...  ...     ...     ...     ...         ...  
3830  0.0  2.5  0.3  0.3  0.4  0.0  0.8   112.1   109.1     3.0   109140000  
3831  0.6  2.1  0.8  0.4  0.0  0.6  0.3   106.8   106.8    -0.1   109140000  
3832  0.6  1.1  2.0  0.4  0.8  0.2 -1.9   107.4   115.8    -8.4   109140000  
3833  0.3  1.0  1.5  0.5  0.2  0.2 -0.5   103.4   107.0    -3.6   109140000  
3834  0.4  1.0  0.5  0.2  0.2  0.0  1.9   109.4    98.4    11.0   109140000  

[3835 rows x 20 columns]

In [34]:
dfPlayerValue.isnull().sum()

PLAYER        0
SALARY        0
SEASON        0
TEAM          0
AGE           0
GP            0
W             0
L             0
PTS           0
3PM           0
REB           0
AST           0
TOV           0
STL           0
BLK           0
+/-           0
OFFRTG        0
DEFRTG        0
NETRTG        0
SALARY_CAP    0
dtype: int64

In [35]:
dfPlayerValue

PLAYER      SALARY  SEASON TEAM   AGE    GP     W     L   PTS  \
0         Kobe Bryant  25244493.0    2010  LAL  32.0  82.0  57.0  25.0  25.3   
1        Vince Carter  21300000.0    2010  PHX  34.0  73.0  38.0  35.0  14.0   
2       Kevin Garnett  21247044.0    2010  BOS  35.0  71.0  49.0  22.0  14.9   
3          Tim Duncan  21164619.0    2010  SAS  35.0  76.0  60.0  16.0  13.4   
4       Rashard Lewis  21136631.0    2010  WAS  31.0  57.0  24.0  33.0  11.7   
...               ...         ...     ...  ...   ...   ...   ...   ...   ...   
3830     Jaylen Hoard    129209.0    2019  POR  21.0  13.0   7.0   6.0   2.9   
3831  Justin Anderson    125000.0    2019  BKN  26.0  10.0   6.0   4.0   2.8   
3832  Jeremiah Martin     58452.0    2019  BKN  24.0   9.0   6.0   3.0   7.1   
3833     Jerian Grant     50000.0    2019  WAS  27.0   6.0   1.0   5.0   4.5   
3834       Louis King     49222.0    2019  DET  21.0  10.0   3.0   7.0   2.0   

      3PM  REB  AST  TOV  STL  BLK  +/-  OFFRTG  DEFRTG  NETRTG  SALARY_CAP  
0     1.4  5.1  4.7  3.0  1.2  0.1  5.2   111.5   104.2     7.3    58044000  
1     1.6  3.8  2.0  1.2  0.9  0.3  1.7   108.0   104.6     3.4    58044000  
2     0.0  8.9  2.4  1.6  1.3  0.8  7.8   109.2    96.4    12.9    58044000  
3     0.0  8.9  2.7  1.6  0.7  1.9  6.0   110.3   100.3    10.0    58044000  
4     1.5  5.1  1.6  1.5  0.9  0.5 -1.2   102.6   104.3    -1.6    58044000  
...   ...  ...  ...  ...  ...  ...  ...     ...     ...     ...         ...  
3830  0.0  2.5  0.3  0.3  0.4  0.0  0.8   112.1   109.1     3.0   109140000  
3831  0.6  2.1  0.8  0.4  0.0  0.6  0.3   106.8   106.8    -0.1   109140000  
3832  0.6  1.1  2.0  0.4  0.8  0.2 -1.9   107.4   115.8    -8.4   109140000  
3833  0.3  1.0  1.5  0.5  0.2  0.2 -0.5   103.4   107.0    -3.6   109140000  
3834  0.4  1.0  0.5  0.2  0.2  0.0  1.9   109.4    98.4    11.0   109140000  

[3835 rows x 20 columns]

In [176]:
# add function to remove the bias caused by agd
df1 = dfPlayerValue.PLAYER.sample(n = 30, replace = False, random_state = 100)
print(sum(dfPlayerValue.PLAYER.isin(df1)))
df2 = dfPlayerValue.loc[dfPlayerValue.PLAYER.isin(df1)]
df2.loc[df2.PLAYER == "Joakim Noah"].SEASON.max()
type(df2.loc[(df2.PLAYER == "Joakim Noah") & (df2.SEASON == 2019)])
df2

183


PLAYER      SALARY  SEASON TEAM   AGE    GP     W     L  \
1          Vince Carter  21300000.0    2010  PHX  34.0  73.0  38.0  35.0   
28         Al Jefferson  14000000.0    2010  UTA  26.0  82.0  39.0  43.0   
47         Marcus Camby  11227250.0    2010  POR  37.0  59.0  32.0  27.0   
54          Rajon Rondo  10045455.0    2010  BOS  25.0  68.0  47.0  21.0   
60    Richard Jefferson   9282000.0    2010  SAS  31.0  81.0  60.0  21.0   
...                 ...         ...     ...  ...   ...   ...   ...   ...   
3598         Mike Scott   5005350.0    2019  PHI  31.0  68.0  39.0  29.0   
3603    Pat Connaughton   4938273.0    2019  MIL  27.0  67.0  53.0  14.0   
3607     Garrett Temple   4767000.0    2019  BKN  34.0  62.0  30.0  32.0   
3659        Alex Caruso   2750000.0    2019  LAL  26.0  64.0  48.0  16.0   
3666      Udonis Haslem   2564753.0    2019  MIA  40.0   4.0   3.0   1.0   

       PTS  3PM   REB   AST  TOV  STL  BLK  +/-  OFFRTG  DEFRTG  NETRTG  \
1     14.0  1.6   3.8   2.0  1.2  0.9  0.3  1.7   108.0   104.6     3.4   
28    18.6  0.0   9.7   1.8  1.3  0.6  1.9 -2.0   107.7   110.8    -3.1   
47     4.7  0.0  10.3   2.1  1.1  0.7  1.6  0.5   106.4   105.1     1.3   
54    10.6  0.1   4.4  11.2  3.4  2.3  0.2  6.5   108.6    99.8     8.9   
60    11.0  1.7   3.8   1.3  1.1  0.5  0.4  3.6   109.3   103.0     6.3   
...    ...  ...   ...   ...  ...  ...  ...  ...     ...     ...     ...   
3598   6.0  1.2   3.6   0.8  0.4  0.3  0.1  2.1   110.2   104.6     5.7   
3603   5.4  0.8   4.2   1.6  0.8  0.4  0.5  1.2   107.3   103.8     3.5   
3607  10.3  2.0   3.5   2.5  1.0  0.8  0.5 -0.7   105.6   106.9    -1.4   
3659   5.5  0.6   1.9   1.9  0.8  1.1  0.3  3.8   110.8   101.0     9.8   
3666   3.0  0.3   4.0   0.3  0.3  0.0  0.0 -4.8    84.3   104.4   -20.2   

      SALARY_CAP  
1       58044000  
28      58044000  
47      58044000  
54      58044000  
60      58044000  
...          ...  
3598   109140000  
3603   109140000  
3607   109140000  
3659   109140000  
3666   109140000  

[183 rows x 20 columns]

In [177]:
df3 = addEliminatedPlayerRecord(df2)
df3[df3.PLAYER == "Marcus Camby"]

/var/folders/2t/5fzjt2x13h329_bxg11ttqzh0000gn/T/ipykernel_34023/1476788958.py:10: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  df = df.append(pd.Series(), ignore_index=True)


PLAYER      SALARY  SEASON TEAM   AGE    GP     W     L  PTS  3PM  \
2    Marcus Camby  11227250.0  2010.0  POR  37.0  59.0  32.0  27.0  4.7  0.0   
29   Marcus Camby   4590338.0  2011.0  HOU  38.0  59.0  28.0  31.0  4.9  0.0   
47   Marcus Camby   3383773.0  2012.0  NYK  39.0  24.0  16.0   8.0  1.8  0.0   
186  Marcus Camby         0.0  2013.0  ELM  40.0   0.0   0.0   0.0  0.0  0.0   
187  Marcus Camby         0.0  2014.0  ELM  41.0   0.0   0.0   0.0  0.0  0.0   
188  Marcus Camby         0.0  2015.0  ELM  42.0   0.0   0.0   0.0  0.0  0.0   
189  Marcus Camby         0.0  2016.0  ELM  43.0   0.0   0.0   0.0  0.0  0.0   
190  Marcus Camby         0.0  2017.0  ELM  44.0   0.0   0.0   0.0  0.0  0.0   
191  Marcus Camby         0.0  2018.0  ELM  45.0   0.0   0.0   0.0  0.0  0.0   
192  Marcus Camby         0.0  2019.0  ELM  46.0   0.0   0.0   0.0  0.0  0.0   

      REB  AST  TOV  STL  BLK  +/-  OFFRTG  DEFRTG  NETRTG   SALARY_CAP  
2    10.3  2.1  1.1  0.7  1.6  0.5   106.4   105.1     1.3   58044000.0  
29    9.0  1.8  1.0  0.8  1.4  0.6   104.3   103.0     1.3   58044000.0  
47    3.3  0.6  0.6  0.3  0.6 -0.7    96.4    98.3    -1.9   58679000.0  
186   0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0   63065000.0  
187   0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0   70000000.0  
188   0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0   94143000.0  
189   0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0   99093000.0  
190   0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0  101869000.0  
191   0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0  109140000.0  
192   0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0  109140000.0

In [171]:
def addEliminatedPlayerRecord(df):
    SEASON_MAX = 2019
    # check all players' names, identify their last season
    for names in df['PLAYER'].unique():
        if (df.loc[df.PLAYER == names].SEASON.max() < SEASON_MAX):
            season_eliminated = (df.loc[df.PLAYER == names].SEASON.max()) + 1
            for ss in range(int(season_eliminated), (SEASON_MAX + 1)):
                # add records to the season of SEASON_MAX with all other fields as 0, 
                # except the age plus 1 per season
                df = df.append(pd.Series(), ignore_index=True)
                
                age = df.loc[df.PLAYER == names].AGE.max() + 1
                df.iloc[-1, 0] = names
                df.iloc[-1, 2] = ss
                df.iloc[-1, 3] = "ELM"
                df.iloc[-1, 4] = age
                df.iloc[-1, -1] = df_s_cap[df_s_cap['SEASON']==ss].SALARY_CAP
                df = df.fillna(0)
    return df

In [178]:
dfPlayerValue_UnbiasedAge = addEliminatedPlayerRecord(dfPlayerValue)
dfPlayerValue_UnbiasedAge

/var/folders/2t/5fzjt2x13h329_bxg11ttqzh0000gn/T/ipykernel_34023/1476788958.py:10: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  df = df.append(pd.Series(), ignore_index=True)


PLAYER      SALARY  SEASON TEAM   AGE    GP     W     L  \
0            Kobe Bryant  25244493.0  2010.0  LAL  32.0  82.0  57.0  25.0   
1           Vince Carter  21300000.0  2010.0  PHX  34.0  73.0  38.0  35.0   
2          Kevin Garnett  21247044.0  2010.0  BOS  35.0  71.0  49.0  22.0   
3             Tim Duncan  21164619.0  2010.0  SAS  35.0  76.0  60.0  16.0   
4          Rashard Lewis  21136631.0  2010.0  WAS  31.0  57.0  24.0  33.0   
...                  ...         ...     ...  ...   ...   ...   ...   ...   
6551        Jaylen Adams         0.0  2019.0  ELM  24.0   0.0   0.0   0.0   
6552         Joe Chealey         0.0  2019.0  ELM  24.0   0.0   0.0   0.0   
6553  Johnathan Williams         0.0  2019.0  ELM  25.0   0.0   0.0   0.0   
6554        Jonah Bolden         0.0  2019.0  ELM  24.0   0.0   0.0   0.0   
6555          BJ Johnson         0.0  2019.0  ELM  24.0   0.0   0.0   0.0   

       PTS  3PM  REB  AST  TOV  STL  BLK  +/-  OFFRTG  DEFRTG  NETRTG  \
0     25.3  1.4  5.1  4.7  3.0  1.2  0.1  5.2   111.5   104.2     7.3   
1     14.0  1.6  3.8  2.0  1.2  0.9  0.3  1.7   108.0   104.6     3.4   
2     14.9  0.0  8.9  2.4  1.6  1.3  0.8  7.8   109.2    96.4    12.9   
3     13.4  0.0  8.9  2.7  1.6  0.7  1.9  6.0   110.3   100.3    10.0   
4     11.7  1.5  5.1  1.6  1.5  0.9  0.5 -1.2   102.6   104.3    -1.6   
...    ...  ...  ...  ...  ...  ...  ...  ...     ...     ...     ...   
6551   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0   
6552   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0   
6553   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0   
6554   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0   
6555   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0     0.0     0.0   

       SALARY_CAP  
0      58044000.0  
1      58044000.0  
2      58044000.0  
3      58044000.0  
4      58044000.0  
...           ...  
6551  109140000.0  
6552  109140000.0  
6553  109140000.0  
6554  109140000.0  
6555  109140000.0  

[6556 rows x 20 columns]

In [182]:
dfPlayerValue_UnbiasedAge.SEASON.value_counts()
dfPlayerValue_UnbiasedAge[dfPlayerValue_UnbiasedAge.PLAYER == "Jaylen Adams"]
dfPlayerValue_UnbiasedAge.isnull().sum()

PLAYER        0
SALARY        0
SEASON        0
TEAM          0
AGE           0
GP            0
W             0
L             0
PTS           0
3PM           0
REB           0
AST           0
TOV           0
STL           0
BLK           0
+/-           0
OFFRTG        0
DEFRTG        0
NETRTG        0
SALARY_CAP    0
dtype: int64

In [181]:
dfPlayerValue_UnbiasedAge.to_csv('training_unbiased.csv', index = False)

In [185]:
dfPlayerValue.L

0       25.0
1       35.0
2       22.0
3       16.0
4       33.0
        ... 
3830     6.0
3831     4.0
3832     3.0
3833     5.0
3834     7.0
Name: L, Length: 3835, dtype: float64

In [36]:
dfPlayerValue.to_csv('training.csv', index=False)

In [37]:
df_s_2021.SALARY = [x[1:] for x in df_s_2021.SALARY]
df_s_2021.SALARY = df_s_2021.SALARY.str.replace(",", "")
df_s_2021['SALARY'] = pd.to_numeric(df_s_2021['SALARY'])
df_s_2021

PLAYER    SALARY  SEASON
0        Stephen Curry  45780966    2021
1         James Harden  44310840    2021
2            John Wall  44310840    2021
3    Russell Westbrook  44211146    2021
4         LeBron James  41180544    2021
..                 ...       ...     ...
531  Admiral Schofield    300000    2021
532      Troy Williams    122741    2021
533     Mamadi Diakite    100000    2021
534  Demetrius Jackson     92857    2021
535  Freddie Gillespie     50000    2021

[536 rows x 3 columns]

In [38]:
df_s_2021.loc[:,'SALARY_CAP'] = 112414000

In [39]:
df_s_2021.loc[:,"SEASON"] = df_s_2021.loc[:,"SEASON"] - 1
df_s_2021

PLAYER    SALARY  SEASON  SALARY_CAP
0        Stephen Curry  45780966    2020   112414000
1         James Harden  44310840    2020   112414000
2            John Wall  44310840    2020   112414000
3    Russell Westbrook  44211146    2020   112414000
4         LeBron James  41180544    2020   112414000
..                 ...       ...     ...         ...
531  Admiral Schofield    300000    2020   112414000
532      Troy Williams    122741    2020   112414000
533     Mamadi Diakite    100000    2020   112414000
534  Demetrius Jackson     92857    2020   112414000
535  Freddie Gillespie     50000    2020   112414000

[536 rows x 4 columns]

In [40]:
df_sta_2021 = pd.merge(df_s_2021, df_ta_2020, on=['PLAYER','SEASON'], how = 'inner')
df_sta_2021


PLAYER    SALARY  SEASON  SALARY_CAP  Unnamed: 0_x TEAM  AGE  \
0        Stephen Curry  45780966    2020   112414000             1  GSW   33   
1         James Harden  44310840    2020   112414000            21  BKN   31   
2            John Wall  44310840    2020   112414000            38  HOU   30   
3    Russell Westbrook  44211146    2020   112414000            30  WAS   32   
4         LeBron James  41180544    2020   112414000            17  LAL   36   
..                 ...       ...     ...         ...           ...  ...  ...   
382    Cassius Winston    462629    2020   112414000           497  WAS   23   
383      Mychal Mulder    449155    2020   112414000           335  GSW   27   
384      Jontay Porter    300000    2020   112414000           488  MEM   21   
385     Mamadi Diakite    100000    2020   112414000           439  MIL   24   
386  Freddie Gillespie     50000    2020   112414000           339  TOR   24   

     GP   W   L  ...  OREB%  DREB%  REB%  TO RATIO  EFG%   TS%  USG%    PACE  \
0    63  37  26  ...    1.3   13.5   7.7      10.2  60.5  65.5  33.1  105.16   
1    44  31  13  ...    2.3   19.1  11.1      11.6  54.8  61.8  28.3   99.78   
2    40  12  28  ...    1.2    8.4   4.7      11.4  45.8  50.3  31.6  101.42   
3    65  30  35  ...    4.3   24.9  14.7      12.5  47.4  50.9  29.5  107.18   
4    45  30  15  ...    2.0   20.2  11.4      11.7  57.6  60.2  31.0  100.41   
..   ..  ..  ..  ...    ...    ...   ...       ...   ...   ...   ...     ...   
382  22  10  12  ...    0.0    8.7   4.4       9.4  54.5  57.5  18.2  102.65   
383  60  33  27  ...    1.0    6.2   3.7       5.0  60.9  61.3  15.7  105.81   
384  11   7   4  ...    6.5   14.9  10.9      28.0  63.3  64.0  18.2  102.24   
385  14   7   7  ...    5.2   16.0  10.7       5.2  41.3  47.7  14.2  107.54   
386  20   7  13  ...   10.0   14.2  12.0       9.9  52.4  56.3  12.0  100.86   

      PIE   POSS  
0    18.1  4,725  
1    17.4  3,347  
2    11.4  2,717  
3    16.4  5,279  
4    19.1  3,145  
..    ...    ...  
382   7.9    208  
383   6.2  1,695  
384   6.7    115  
385   8.1    314  
386   7.5    824  

[387 rows x 50 columns]

In [41]:
df_sta_2021.columns
df_sta_2021.isnull().sum()
predictors = ['PLAYER', 'SALARY', 'SEASON', 'TEAM', 'AGE', 'GP', 'W', 'L', \
              'PTS', '3PM', 'REB', 'AST', 'TOV', 'STL', 'BLK', '+/-', \
              'OFFRTG', 'DEFRTG', 'NETRTG']
dfPlayerValue_test = df_sta_2021.loc[:, predictors]
dfPlayerValue_test.loc[:,"SALARY_CAP"] = 112414000
dfPlayerValue_test

PLAYER    SALARY  SEASON TEAM  AGE  GP   W   L   PTS  3PM  \
0        Stephen Curry  45780966    2020  GSW   33  63  37  26  32.0  5.3   
1         James Harden  44310840    2020  BKN   31  44  31  13  24.6  2.8   
2            John Wall  44310840    2020  HOU   30  40  12  28  20.6  2.0   
3    Russell Westbrook  44211146    2020  WAS   32  65  30  35  22.2  1.3   
4         LeBron James  41180544    2020  LAL   36  45  30  15  25.0  2.3   
..                 ...       ...     ...  ...  ...  ..  ..  ..   ...  ...   
382    Cassius Winston    462629    2020  WAS   23  22  10  12   1.9  0.4   
383      Mychal Mulder    449155    2020  GSW   27  60  33  27   5.6  1.4   
384      Jontay Porter    300000    2020  MEM   21  11   7   4   2.0  0.3   
385     Mamadi Diakite    100000    2020  MIL   24  14   7   7   3.1  0.1   
386  Freddie Gillespie     50000    2020  TOR   24  20   7  13   5.6  0.0   

      REB   AST  TOV  STL  BLK  +/-  OFFRTG  DEFRTG  NETRTG  SALARY_CAP  
0     5.5   5.8  3.4  1.2  0.1  3.9   114.2   109.6     4.6   112414000  
1     7.9  10.8  4.0  1.2  0.8  2.3   117.2   114.3     2.9   112414000  
2     3.2   6.9  3.5  1.1  0.8 -4.2   106.1   111.8    -5.7   112414000  
3    11.5  11.7  4.8  1.4  0.4 -1.4   109.2   110.4    -1.2   112414000  
4     7.7   7.8  3.7  1.1  0.6  6.4   112.4   103.2     9.2   112414000  
..    ...   ...  ...  ...  ...  ...     ...     ...     ...         ...  
382   0.4   0.5  0.2  0.1  0.0 -2.2    92.8   114.2   -21.4   112414000  
383   1.0   0.4  0.3  0.2  0.2  1.1   111.3   108.4     2.9   112414000  
384   1.3   0.1  0.6  0.3  0.1  0.1    93.0    92.2     0.9   112414000  
385   2.4   0.6  0.2  0.4  0.4  0.6   103.2    99.4     3.8   112414000  
386   4.9   0.5  0.6  0.7  1.0 -3.9   101.3   110.9    -9.6   112414000  

[387 rows x 20 columns]

In [42]:
dfPlayerValue_test.to_csv('testing.csv', index = False)

### Michael's suggestion
- many variable: MLR first to select variable, then HMLR
- Direct HMLR applied to u only clear on ur predictors are
- influence.me for diagnose for HMLR model

- there will be cap limit for each player at max, e.g. 35% (check that)
- include all star appearance in predictors